### PR TITLE
Add garbage collection function for segments produced by partially failed imports

### PR DESCRIPTION
**Pull request by EricStarer**
Add garbage collection function for segments produced by partially failed imports

Signed-off-by: xingzhao <xing.zhao@zilliz.com>

https://github.com/milvus-io/milvus/issues/17814

/kind improvement

### CODE REVIEW

File: internal/datacoord/garbage_collector.go [hunk 1]
```diff
@@ -158,8 +163,28 @@ func (gc *garbageCollector) scan() {
...
+	failedSegmentIDs := rsp.SegmentIDs
+	failedSegmentIDsSet := make(map[int64]struct{})
+	for _, failedSegmentID := range failedSegmentIDs {
+		failedSegmentIDsSet[failedSegmentID] = struct{}{}
+	}
+
 	drops := gc.meta.SelectSegments(func(segment *SegmentInfo) bool {
-		return segment.GetState() == commonpb.SegmentState_Dropped && !gc.segRefer.HasSegmentLock(segment.ID)
+		_, failedSegmentFlag := failedSegmentIDsSet[segment.ID]
+		if failedSegmentFlag {
```
Comment (soothing-rain) [inline internal/datacoord/garbage_collector.go:184]: This seems duplicated as the flag is checked in at line #187

File: internal/datacoord/garbage_collector.go [hunk 2]
```diff
@@ -158,8 +163,28 @@ func (gc *garbageCollector) scan() {
 }
 
 func (gc *garbageCollector) clearEtcd() {
+	req := &internalpb.GetImportFailedSegmentIDsRequest{
+		Base: &commonpb.MsgBase{
+			MsgType: commonpb.MsgType_GetImportFailedSegmentIDs,
+		},
+	}
+	ctx := context.Background()
```
Comment (soothing-rain) [inline internal/datacoord/garbage_collector.go:171]: Check the healthiness of rootcoord before sending RPC

File: internal/datacoord/garbage_collector.go [hunk 3]
```diff
@@ -54,21 +56,24 @@ type garbageCollector struct {
 	meta     *meta
 	segRefer *SegmentReferenceManager
 
+	rootCoordClient types.RootCoord
```
Comment (soothing-rain) [inline internal/datacoord/garbage_collector.go:59]: Go prefers shorter variable names, for example `rcc`.

File: internal/datacoord/garbage_collector.go [hunk 4]
```diff
@@ -158,8 +163,28 @@ func (gc *garbageCollector) scan() {
...
+		Base: &commonpb.MsgBase{
+			MsgType: commonpb.MsgType_GetImportFailedSegmentIDs,
+		},
+	}
+	ctx := context.Background()
+	rsp, err := gc.rootCoordClient.GetImportFailedSegmentIDs(ctx, req)
+	if err != nil {
+		log.Error("getImportFailedSegmentIDs failed", zap.String("error", err.Error()))
+	}
+	failedSegmentIDs := rsp.SegmentIDs
```
Comment (soothing-rain) [inline internal/datacoord/garbage_collector.go:176]: Here and below, use "rsp.GetSegmentIDs()" instead.

File: internal/rootcoord/root_coord.go [hunk 5]
```diff
@@ -3021,3 +3021,16 @@ func (c *Core) ListCredUsers(ctx context.Context, in *milvuspb.ListCredUsersRequ
 		Usernames: credInfo.Usernames,
 	}, nil
 }
+
+func (c *Core) GetImportFailedSegmentIDs(ctx context.Context, req *internalpb.GetImportFailedSegmentIDsRequest) (*internalpb.GetImportFailedSegmentIDsResponse, error) {
```
Comment (soothing-rain) [inline internal/rootcoord/root_coord.go:3025]: Comment

File: internal/rootcoord/import_manager.go [hunk 6]
```diff
@@ -729,3 +729,18 @@ func taskExpired(ti *datapb.ImportTaskInfo) bool {
 		ti.GetState().GetStateCode() != commonpb.ImportState_ImportCompleted &&
 		Params.RootCoordCfg.ImportTaskExpiration <= float64(time.Now().Unix()-ti.GetCreateTs())
 }
+
+func (m *importManager) GetImportFailedSegmentIDs() ([]int64, error) {
```
Comment (soothing-rain) [inline internal/rootcoord/import_manager.go:733]: comment

File: internal/rootcoord/import_manager.go [hunk 7]
```diff
@@ -729,3 +729,18 @@ func taskExpired(ti *datapb.ImportTaskInfo) bool {
...
+	ret := make([]int64, 0, 10)
```
Comment (soothing-rain) [inline internal/rootcoord/import_manager.go:734]: Why is there a cap of 10 and what happens if it is exceeded?

File: internal/rootcoord/import_manager.go [hunk 8]
```diff
@@ -729,3 +729,18 @@ func taskExpired(ti *datapb.ImportTaskInfo) bool {
...
+	for _, importTaskInfo := range m.pendingTasks {
```
Comment (soothing-rain) [inline internal/rootcoord/import_manager.go:735]: locks are required while reading.

File: internal/datacoord/garbage_collector.go (hunk 3 above)
Comment (EricStarer) [reply internal/datacoord/garbage_collector.go:59]: ok

File: internal/datacoord/garbage_collector.go (hunk 2 above)
Comment (EricStarer) [reply internal/datacoord/garbage_collector.go:171]: this may be checked at rootcoord server

Comment (soothing-rain) [reply internal/datacoord/garbage_collector.go:171]: Take a look at: https://github.com/milvus-io/milvus/blob/f0b036a35ae17ada3101eebc6a200b5f621b7e7e/internal/rootcoord/root_coord.go#L216

Comment (EricStarer) [reply internal/datacoord/garbage_collector.go:171]: this method is private and can only be call at rootcoord package？

File: internal/rootcoord/import_manager.go (hunk 6 above)
Comment (EricStarer) [reply internal/rootcoord/import_manager.go:733]: what does this mean?

File: internal/rootcoord/import_manager.go (hunk 7 above)
Comment (EricStarer) [reply internal/rootcoord/import_manager.go:734]: If it's exceeded, the slice will be copy to a new memory space and enlarge its capacity.

File: internal/datacoord/garbage_collector.go (hunk 1 above)
Comment (EricStarer) [reply internal/datacoord/garbage_collector.go:185]: I will first set the segment state into dropped to make it easier to clear in the next time if the sevice is crashed.

File: internal/rootcoord/import_manager.go (hunk 7 above)
Comment (soothing-rain) [reply internal/rootcoord/import_manager.go:734]: If we not going to limit the size, we can just do: `ret := make([]int64, 0)`.

File: internal/rootcoord/import_manager.go (hunk 6 above)
Comment (soothing-rain) [reply internal/rootcoord/import_manager.go:733]: add comment for this method. But it seems that you commented in the interface so nvm.

File: internal/datacoord/garbage_collector.go (hunk 2 above)
Comment (soothing-rain) [reply internal/datacoord/garbage_collector.go:171]: OK I guess just checking `rcc != nil` for now

File: internal/rootcoord/import_manager.go (hunk 7 above)
Comment (EricStarer) [reply internal/rootcoord/import_manager.go:734]: ok

File: internal/datacoord/garbage_collector.go (hunk 1 above)
Comment (soothing-rain) [reply internal/datacoord/garbage_collector.go:185]: That makes sense. Then how about just remove the flag check in the next line?

File: internal/rootcoord/import_manager.go (hunk 6 above)
Comment (EricStarer) [reply internal/rootcoord/import_manager.go:733]: ok

File: internal/datacoord/garbage_collector.go (hunk 2 above)
Comment (EricStarer) [reply internal/datacoord/garbage_collector.go:171]: ok

File: internal/datacoord/garbage_collector.go (hunk 1 above)
Comment (EricStarer) [reply internal/datacoord/garbage_collector.go:189]: ok

File: internal/datacoord/garbage_collector.go [hunk 9]
```diff
@@ -158,7 +163,33 @@ func (gc *garbageCollector) scan() {
 }
 
 func (gc *garbageCollector) clearEtcd() {
+	failedSegmentIDs := make([]int64, 0)
+	if gc.rcc != nil {
```
Comment (soothing-rain) [inline internal/datacoord/garbage_collector.go:167]: Log error and return if rcc == nil

Comment (EricStarer) [reply internal/datacoord/garbage_collector.go:167]: ok

Comment (EricStarer) [reply internal/datacoord/garbage_collector.go:167]: Maybe we can log warn when rcc == nil.

**Comment by EricStarer** [comment]
rerun ut

**Comment by EricStarer** [comment]
rerun ut
/run-checks

**Comment by EricStarer** [comment]
rerun ut

**Comment by EricStarer** [comment]
rerun ut

File: internal/datacoord/garbage_collector.go [hunk 10]
```diff
@@ -158,7 +163,35 @@ func (gc *garbageCollector) scan() {
...
+
+	failedSegmentIDsSet := make(map[int64]struct{})
+	for _, failedSegmentID := range failedSegmentIDs {
+		failedSegmentIDsSet[failedSegmentID] = struct{}{}
+	}
+
 	drops := gc.meta.SelectSegments(func(segment *SegmentInfo) bool {
+		_, failedSegmentFlag := failedSegmentIDsSet[segment.ID]
+		if failedSegmentFlag {
+			segment.State = commonpb.SegmentState_Dropped
```
Comment (soothing-rain) [inline internal/datacoord/garbage_collector.go:192]: One more thing, do not update segment fields directly. Use copy on write:

https://github.com/milvus-io/milvus/blob/899a9bf86d16bc45422858fa2fff193f043b7c89/internal/datacoord/segment_info.go#L103-L107

Comment (EricStarer) [reply internal/datacoord/garbage_collector.go]: done

**Pull request closed without merging**